### PR TITLE
Implement T::Enum#=== intrinsic

### LINKED ISSUE #4115: A case over T::Enum should not require an explicit T.absurd
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0Aextend%20T%3A%3ASig%0A%0Aclass%20MyEnum%20%3C%20T%3A%3AEnum%0A%20%20enums%20do%0A%20%20%20%20X%20%3D%20new%0A%20%20%20%20Y%20%3D%20new%0A%20%20end%0Aend%0A%0Asig%20%7Bparams(x%3A%20MyEnum).returns(Integer)%7D%0Adef%20example(x)%0A%20%20res%20%3D%20case%20x%0A%20%20when%20MyEnum%3A%3AX%20then%201%0A%20%20when%20MyEnum%3A%3AY%20then%202%0A%20%20end%0A%0A%20%20res%0Aend)

```ruby
# typed: strict
extend T::Sig

class MyEnum < T::Enum
  enums do
    X = new
    Y = new
  end
end

sig {params(x: MyEnum).returns(Integer)}
def example(x)
  res = case x
  when MyEnum::X then 1
  when MyEnum::Y then 2
  end

  res
end
```

#### Observed output

```
editor.rb:18: Expected Integer but found T.nilable(Integer) for method result type https://srb.help/7005
    18 |  res
          ^^^
  Expected Integer for result type of method example:
    editor.rb:12:
    12 |def example(x)
        ^^^^^^^^^^^^^^
  Got T.nilable(Integer) originating from:
    editor.rb:14:
    14 |  when MyEnum::X then 1
                              ^
    editor.rb:15:
    15 |  when MyEnum::Y then 2
                              ^
    editor.rb:12: Possibly uninitialized (NilClass) in:
    12 |def example(x)
        ^^^^^^^^^^^^^^
Errors: 1
```

#### Expected behavior

There should be no error, because the `case` is already exhaustive, so Sorbet should know that the (implicit) `else` branch of the `case` is dead and thus `res` is always initialized with an `Integer`.

- - -

By way of comparison, if we make an analogous example using explicit sealed classes instead of `T::Enum`, the `T.absurd` is not required:

```ruby
# typed: strict
extend T::Sig

class MyEnum
  extend T::Helpers
  abstract!
  sealed!

  class X < MyEnum; include Singleton; final!; end
  class Y < MyEnum; include Singleton; final!; end
end

sig {params(x: MyEnum).returns(Integer)}
def example_with_class_objects(x)
  res = case x
  when MyEnum::X then 1
  when MyEnum::Y then 2
  end

  res # no error
end
```

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0Aextend%20T%3A%3ASig%0A%0Aclass%20MyEnum%0A%20%20extend%20T%3A%3AHelpers%0A%20%20abstract!%0A%20%20sealed!%0A%0A%20%20class%20X%20%3C%20MyEnum%3B%20include%20Singleton%3B%20final!%3B%20end%0A%20%20class%20Y%20%3C%20MyEnum%3B%20include%20Singleton%3B%20final!%3B%20end%0Aend%0A%0Asig%20%7Bparams(x%3A%20MyEnum).returns(Integer)%7D%0Adef%20example_with_class_objects(x)%0A%20%20res%20%3D%20case%20x%0A%20%20when%20MyEnum%3A%3AX%20then%201%0A%20%20when%20MyEnum%3A%3AY%20then%202%0A%20%20end%0A%0A%20%20res%0Aend)

And another example: if instead of the class objects (`MyEnum::X`) we use the singleton instance objects (`MyEnum::X.instance`), we see the same behavior as the example using `T::Enum`:

```ruby
# typed: strict
extend T::Sig

class MyEnum
  extend T::Helpers
  abstract!
  sealed!

  class X < MyEnum; include Singleton; final!; end
  class Y < MyEnum; include Singleton; final!; end
end

sig {params(x: MyEnum).returns(Integer)}
def example_with_singleton_instances(x)
  res = case x
  when MyEnum::X.instance then 1
  when MyEnum::Y.instance then 2
  end

  res # error: Expected `Integer` but found `T.nilable(Integer)` for method result type
end
```

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0Aextend%20T%3A%3ASig%0A%0Aclass%20MyEnum%0A%20%20extend%20T%3A%3AHelpers%0A%20%20abstract!%0A%20%20sealed!%0A%0A%20%20class%20X%20%3C%20MyEnum%3B%20include%20Singleton%3B%20final!%3B%20end%0A%20%20class%20Y%20%3C%20MyEnum%3B%20include%20Singleton%3B%20final!%3B%20end%0Aend%0A%0Asig%20%7Bparams(x%3A%20MyEnum).returns(Integer)%7D%0Adef%20example_with_singleton_instances(x)%0A%20%20res%20%3D%20case%20x%0A%20%20when%20MyEnum%3A%3AX.instance%20then%201%0A%20%20when%20MyEnum%3A%3AY.instance%20then%202%0A%20%20end%0A%0A%20%20res%0Aend)

### PR DESCRIPTION
**Pull request by aprocter-stripe**
The root of #4115 seems to be that the `case` desugars to `===` on the singletons for the enums, not the classes. Implementing an intrinsic for `T::Enum#===` seems to me like a reasonable way to fix it, ~~but I'm tossing a wip PR up first to get feedback on whether it's the right approach~~.

### Motivation
Fixes #4115

### Test plan
See included automated tests.

### CODE REVIEW

**Comment by aprocter-stripe** [comment]
(Just for some additional context: this does seem to fix the errors in #4115, and I think it matches the behavior of the runtime [see https://github.com/sorbet/sorbet/blob/b12cbff755c13f2c659ccfa4688e78729d8faa46/gems/sorbet-runtime/lib/types/enum.rb#L220-L233] except for the `legacy_t_enum_migration_mode`.)

File: core/types/calls.cc
Comment (aprocter-stripe) [inline core/types/calls.cc:3013]: Note that the implementation is based on that of `Module_tripleEq` above.

Comment (elliottt) [reply core/types/calls.cc:3013]: Oh, `Module_tripleEq` wasn't firing because the receiver was a value of one of the generated enum classes?

Comment (aprocter-stripe) [reply core/types/calls.cc:3013]: Yup!

**Review by elliottt — CHANGES_REQUESTED** [review]
I think that this approach makes sense, given that the receiver of `===` before was one of the singleton enum values instead of a class/module. What do you think @jez?

File: core/types/calls.cc
Comment (jez) [inline core/types/calls.cc:3031]: Can you make sure to write a test for this?

I think that this should work:

```ruby
T.reveal_type(MyEnum::Y.===(MyEnum::X))
```

also it would be great to have both the examples from #4115 committed as test cases.

**Comment by jez** [comment]
> I think that this approach makes sense, given that the receiver of `===` before was one of the singleton enum values instead of a class/module. What do you think @jez?

This is the best approach. Thanks!

**Review by elliottt — APPROVED** [review]
Nice, great tests!

**Comment by aprocter-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JqWM7IQSVLNGVZ
→ https://go/builds/bui_JqWNbeFJ6YHUR4

**Comment by aprocter-stripe** [comment]
Looks like I goofed something up here.

```ruby
# typed: true
extend T::Sig

class MyEnum < T::Enum ; end

sig {params(x: T::Enum, y: MyEnum).void}
def f(x, y)
  T.reveal_type(x===y)
end

sig {params(x: MyEnum, y: MyEnum).void}
def g(x, y)
  T.reveal_type(x===y)
end
```

On sorbet.run, both of the above reveal as `T::Boolean`. On this branch they reveal as `T::TrueType`, which is uh wrong. :) Will look into what I did wrong in cargo-culting from `Module#===`.

**Comment by aprocter-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JqcojtMIA1LKLN
→ https://go/builds/bui_JqcoNstf71fgLx

**Comment by aprocter-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JqdJt09NHtwwq7
→ https://go/builds/bui_JqdJ8CzkfQ6QI5

**Comment by aprocter-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JqfS5Hym0W3K8i
→ https://go/builds/bui_JqfSnnD4hB2MoH

**Comment by aprocter-stripe** [comment]
So I fixed a couple things:

1. Unbroke behavior when `===`-comparing abstract enum things (i.e. `x===y` where x and y have `T::Enum` or `SomeEnum`; initially the issues here didn't reveal themselves since I was just looking at desugared `case` statements).
2. Added support for `String` on the rhs, to support `legacy_t_enum_migration_mode` (just returns `T::Boolean`).

I'm feeling reasonably good about the correctness here, and I have a green build on Stripe's codebase that doesn't require anything surprising, but since the logic has changed quite a bit a second review is probably in order.

File: core/types/calls.cc
Comment (elliottt) [inline core/types/calls.cc:3087]:
```suggestion
// Returns true if the referenced class is a grandchild of T::Enum, i.e., one of the classes for the singleton enum
// values.
static bool isEnumValueClass(const GlobalState &gs, ClassOrModuleRef klass) {
```

File: core/types/calls.cc
Comment (elliottt) [inline core/types/calls.cc:3119]:
```suggestion
        if (Types::isSubType(gs, rhs, lhs) && isEnumValueClass(gs, lhs_unwrapped) && isEnumValueClass(gs, rhs_unwrapped)) {
```

Comment (elliottt) [reply core/types/calls.cc:3119]: What do you think about reordering this conditional to check if rhs is a subtype of lhs last?

Comment (elliottt) [reply core/types/calls.cc:3119]: What about rewriting this to check that RHS is a subtype of LHS, and LHS is a ClassType with a TEnum name? (You can use the `NameRef::isTEnumName` helper to test that.)

**Review by elliottt — CHANGES_REQUESTED** [review]

**Comment by aprocter-stripe** [comment]
@elliottt Think I got everything, and also added a test for "subset" enums as you pointed out in Slack. (Those actually did seem to work with the old approach (checking for grandchildren of T::Enum), and they also work with the new (checking the name kind).)

**Review by elliottt — APPROVED** [review]
I think this looks great, thanks for all the tests!

File: core/types/calls.cc
Comment (aprocter-stripe) [reply core/types/calls.cc:3096]: Good call --- I think inlining is the way to go. (Started writing it this way 'cause I thought I'd need it in more than one place.) Will do that before merge.

**Comment by aprocter-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JqxlF7P4JZZNMR
→ https://go/builds/bui_JqxmI6BrAFpFVW

**Comment by aprocter-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JrJaYE9CYrk26d
→ https://go/builds/bui_JrJaP2T0kFNkSY

**Pull request merged**
